domain/secret/state/state.go

```diff
@@ -2475,9 +2503,174 @@ AND  (
 	return revisionResult, nil
 }
 
-// InitialWatchStatementForObsoleteRevision returns the initial watch statement and the table name for watching obsolete revisions.
+type dbrevisionUUIDs []revisionUUID
+
+// InitialWatchStatementForConsumedSecretsChange returns the initial watch statement and the table name for watching consumed secrets.
+func (st State) InitialWatchStatementForConsumedSecretsChange(unitName string) (string, eventsource.NamespaceQuery) {
+	queryFunc := func(ctx context.Context, runner database.TxnRunner) ([]string, error) {
+		var revUUIDs dbrevisionUUIDs
+		if err := st.getConsumedSecretURIsWithChanges(ctx, runner,
+			"DISTINCT sr.uuid AS &revisionUUID.uuid", revisionUUID{}, &revUUIDs,
+			unitName,
+		); err != nil {
+			return nil, errors.Trace(err)
+		}
+		result := make([]string, len(revUUIDs))
+		for i, rev := range revUUIDs {
+			result[i] = rev.UUID
+		}
+		return result, nil
+	}
+	return "secret_revision", queryFunc
+}
+
+// GetConsumedSecretURIsWithChanges returns the URIs of the secrets consumed by the specified unit that has new revisions.
+func (st State) GetConsumedSecretURIsWithChanges(ctx context.Context, unitName string, revisionIDs ...string) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var dbConsumers secretUnitConsumers
+	if err = st.getConsumedSecretURIsWithChanges(ctx, db,
+		"DISTINCT suc.secret_id AS &secretUnitConsumer.secret_id", secretUnitConsumer{}, &dbConsumers,
+		unitName, revisionIDs...,
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+	secretURIs := make([]string, len(dbConsumers))
+	for i, consumer := range dbConsumers {
+		uri, err := coresecrets.ParseURI(consumer.SecretID)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		secretURIs[i] = uri.String()
+	}
+	return secretURIs, nil
+}
+
+func (st State) getConsumedSecretURIsWithChanges(
```

Comment:
There's no need for this function. Just use bespoke code in the initial watch and get consumed methods. One passes in revision ids, the other doesn't; it's just messy. It's all just prepare a statement and run the query. It can be done in 2 methods, one for each use case.